baselines/ppo2/policies.py

```diff
@@ -0,0 +1,207 @@
+                    value_network_ = policy_network
+                else:
+                    assert callable(value_network)
+                    value_network_ = value_network
+
+                if is_rnn_network(value_network_):
+                    value_state, value_network_ = value_network_(encoded_x, dones)
+
+            if policy_state or value_state:
+                states_list = [state for state in [policy_state, value_state] if state]
+                states = tf.concat(states_list, axis=1)
+                state_placeholder = tf.placeholder_with_default(states, states.get_shape())
+                index = 0
+                for state in states_list:
+                    assert state.get_shape().ndims == 2
+                    size = int(state.get_shape()[1])
+                    state_map[state] = state_placeholder[:, index:index + size]
+                    index += size
+
+        with tf.variable_scope('policy_latent', reuse=tf.AUTO_REUSE):
+            if is_rnn_network(policy_network_):
+                policy_latent, next_policy_state = \
+                    policy_network_(encoded_x, dones, state_map[policy_state])
+                next_states_list.append(next_policy_state)
+            else:
+                policy_latent = policy_network_(encoded_x)
+
+        with tf.variable_scope('value_latent', reuse=tf.AUTO_REUSE):
+            if value_network_ == 'shared':
+                value_latent = policy_latent
+            elif is_rnn_network(value_network_):
+                value_latent, next_value_state = \
+                    value_network_(encoded_x, dones, state_map[value_state])
+                next_states_list.append(next_value_state)
+            else:
+                value_latent = value_network_(encoded_x)
+
+        with tf.name_scope("save_rnn_memory"):
+            if policy_state or value_state:
+                next_states = tf.concat(next_states_list, axis=1)
+                state_info = {'current': state_placeholder,
+                              'next': next_states, }
+                update_op = []
+                index = 0
+                for state in [state for state in [policy_state, value_state] if state]:
+                    size = int(state.get_shape()[1])
+                    update_op.append(state.assign(next_states[:, index:index + size]))
+                    index += size
+                update_op = tf.group(update_op)
+            else:
+                state_info = None
+                update_op = tf.no_op()
+
+        with tf.control_dependencies([update_op]):
+            policy = PolicyWithValue(
+                env=env,
+                observations=X,
+                dones=dones,
+                latent=policy_latent,
+                vf_latent=value_latent,
+                states=state_info,
+                sess=sess,
+                estimate_q=estimate_q,
+            )
+        return policy
+
+    return policy_fn
+
+
+def is_rnn_network(network):
```

Comment:
alright, this is hacky as f***. How about instead we a registration flag that indicates that policy is recurrent to the "register" decorator?

Reply:
Unfortunately, `is_rnn_network` is a translation of the below code in `common.policies.build_policy`. 
To apply  the idea, refactoring of `@register` will be preceded.

```python baselines.common.build_policy
        with tf.variable_scope('pi', reuse=tf.AUTO_REUSE):
            policy_latent = policy_network(encoded_x)
            if isinstance(policy_latent, tuple):
                policy_latent, recurrent_tensors = policy_latent

                if recurrent_tensors is not None:
                    # recurrent architecture, need a few more steps
                    nenv = nbatch // nsteps
                    assert nenv > 0, 'Bad input for recurrent policy: batch size {} smaller than nsteps {}'.format(nbatch, nsteps)
                    policy_latent, recurrent_tensors = policy_network(encoded_x, nenv)
                    extra_tensors.update(recurrent_tensors)
```

Reply:
May I guide your attention to this related issue in the stable baselines: hill-a/stable-baselines#241? Here we discuss the same problem of how to flag a policy as recurrent.

Reply:
Instead of `is_rnn_network()`, I refactored `@register` decorator.
(e6f0d98, b6e6c52)

Reply:
Cool!